File: FastMines_android/src/main/java/fmg/android/app/presentation/MosaicDsViewModel.java
package fmg.android.app.presentation;

import androidx.lifecycle.ViewModel;

import java.util.function.Consumer;

import fmg.android.app.model.dataSource.MosaicDataSource;
import fmg.common.geom.SizeDouble;

/** ViewModel for {@link fmg.android.app.SelectMosaicFragment} */
public class MosaicDsViewModel extends ViewModel {

    private final MosaicDataSource mosaicDS = new MosaicDataSource();
    private Consumer<String> changedMosaicDSCallback;

    public MosaicDsViewModel() {
        mosaicDS.setListener(this::onMosaicDsPropertyChanged);
    }

    public MosaicDataSource getMosaicDS() { return mosaicDS; }

    public SizeDouble getImageSize()         { return mosaicDS.getImageSize(); }
    public void       setImageSize(SizeDouble size) { mosaicDS.setImageSize(size); }

    private void onMosaicDsPropertyChanged(String propertyName) {
        switch (propertyName) {
        case MosaicDataSource.PROPERTY_IMAGE_SIZE:
            // TODO ! notify parent container
            //notifier.firePropertyChanged<SizeDouble>(ev, nameof(this.ImageSize));
            break;
        }
        if (changedMosaicDSCallback != null)
            changedMosaicDSCallback.accept(propertyName);
    }

    @Override
    protected void onCleared() {
        mosaicDS.setListener(null);
        mosaicDS.close();
        changedMosaicDSCallback = null;
    }

    public void setMosaicDSListener(Consumer<String> callback) {
        if (callback == null) {
            // unset
            changedMosaicDSCallback = null;
        } else {
            // set
            if (changedMosaicDSCallback != null)
                throw new IllegalArgumentException("The callback is already set");
            changedMosaicDSCallback = callback;
        }
    }

}

File: FastMines_uwp/FastMines.Uwp.Main/Presentation/MosaicDsViewModel.cs
using System;
using System.ComponentModel;
using fmg.common.notifier;
using fmg.DataModel.DataSources;
using fmg.common.geom;

namespace fmg.common {

    /// <summary> ViewModel for <see cref="SelectMosaicPage"/> </summary>
    public class MosaicsViewModel : INotifyPropertyChanged, IDisposable {

        private readonly MosaicDataSource mosaicDS = new MosaicDataSource();
        protected bool Disposed { get; private set; }
        public event PropertyChangedEventHandler PropertyChanged {
            add    { _notifier.PropertyChanged += value;  }
            remove { _notifier.PropertyChanged -= value;  }
        }
        protected readonly NotifyPropertyChanged _notifier;

        public MosaicsViewModel() {
            mosaicDS.PropertyChanged += OnMosaicDsPropertyChanged;
            _notifier = new NotifyPropertyChanged(this, false);
        }

        public MosaicDataSource MosaicDS => mosaicDS;

        public SizeDouble ImageSize {
            get { return mosaicDS.ImageSize; }
            set { mosaicDS.ImageSize = value; }
        }

        private void OnMosaicDsPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            _notifier.FirePropertyChanged(nameof(this.MosaicDS));
            if (ev.PropertyName == nameof(MosaicDataSource.ImageSize)) {
                // ! notify parent container
                _notifier.FirePropertyChanged<SizeDouble>(ev, nameof(this.ImageSize));
            }
        }

        public void Dispose() {
            if (Disposed)
                return;
            Disposed = true;

            mosaicDS.PropertyChanged -= OnMosaicDsPropertyChanged;
            mosaicDS.Dispose();

            _notifier.Dispose();

            GC.SuppressFinalize(this);
        }

    }

}
